public/app/plugins/panel/datagrid/DataGridPanel.tsx

```diff
@@ -0,0 +1,456 @@
+  DatagridContextMenuData,
+  deleteRows,
+  EMPTY_CELL,
+  EMPTY_GRID_SELECTION,
+  getCellWidth,
+  isDatagridEditEnabled,
+  publishSnapshot,
+  RIGHT_ELEMENT_PROPS,
+  TRAILING_ROW_OPTIONS,
+} from './utils';
+
+interface Props extends PanelProps<PanelOptions> {}
+
+export const DataGridPanel: React.FC<Props> = ({ options, data, id, fieldConfig }) => {
+  const [columnWidths, setColumnWidths] = useState<Map<number, number>>(new Map());
+  const [columns, setColumns] = useState<SizedGridColumn[]>([]);
+  const [contextMenuData, setContextMenuData] = useState<DatagridContextMenuData>({ isContextMenuOpen: false });
+  const [renameColumnInputData, setRenameColumnInputData] = useState<RenameColumnInputData>({ isInputOpen: false });
+  const [gridSelection, setGridSelection] = useState<GridSelection>(EMPTY_GRID_SELECTION);
+  const [columnFreezeIndex, setColumnFreezeIndex] = useState<number>(0);
+  const [toggleSearch, setToggleSearch] = useState<boolean>(false);
+  const [isResizeInProgress, setIsResizeInProgress] = useState<boolean>(false);
+
+  const frame = data.series[options.selectedSeries ?? 0];
+
+  const theme = useTheme2();
+  const gridTheme = {
+    accentColor: theme.colors.primary.main,
+    accentFg: theme.colors.secondary.main,
+    textDark: theme.colors.text.primary,
+    textMedium: theme.colors.text.primary,
+    textLight: theme.colors.text.primary,
+    textBubble: theme.colors.text.primary,
+    textHeader: theme.colors.text.primary,
+    bgCell: theme.colors.background.primary,
+    bgCellMedium: theme.colors.background.primary,
+    bgHeader: theme.colors.background.secondary,
+    bgHeaderHasFocus: theme.colors.background.secondary,
+    bgHeaderHovered: theme.colors.background.secondary,
+  };
+
+  const setGridColumns = useCallback(() => {
+    const typeToIconMap: Map<string, GridColumnIcon> = new Map([
+      [FieldType.number, GridColumnIcon.HeaderNumber],
+      [FieldType.string, GridColumnIcon.HeaderTextTemplate],
+      [FieldType.boolean, GridColumnIcon.HeaderBoolean],
+      [FieldType.time, GridColumnIcon.HeaderDate],
+      [FieldType.other, GridColumnIcon.HeaderReference],
+    ]);
+
+    if (!frame) {
+      return;
+    }
+
+    setColumns([
+      ...frame.fields.map((f, i) => {
+        const displayName = getFieldDisplayName(f, frame);
+        const width = columnWidths.get(i) ?? getCellWidth(f);
+        return {
+          title: displayName,
+          width: width,
+          icon: typeToIconMap.get(f.type),
+          hasMenu: isDatagridEditEnabled(),
+          trailingRowOptions: { targetColumn: --i },
+        };
+      }),
+    ]);
+  }, [columnWidths, frame]);
+
+  useEffect(() => {
+    setGridColumns();
+  }, [frame, setGridColumns]);
+
+  const getCellContent = ([col, row]: Item): GridCell => {
+    const field: Field = frame.fields[col];
+
+    if (!field) {
+      return EMPTY_CELL;
+    }
+
+    const value = field.values.get(row);
+
+    switch (field.type) {
```

Comment:
Have you considered using an object here instead? iirc switches run linearly, and object key lookups execute in constant time. Plus, I think it'd be easier to test/maintain? Just a thought.

Reply:
interesting, would be curious to see some benchmarks showing this.

Reply:
> I'm surprised!

that's the lesson here. any benchmark that's > 6mo old has to be re-evaluated. JITs improve all the time. i've seen a lot :exploding_head: code because some old benchmark showed that the straightforward code was slower 2yrs ago.

Reply:
actually i think the slowness here is mostly from the lookup version doing function calls, which will always add overhead vs basic constructs like `switch` and `for` (vs forEach)